ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/damle/SandboxDamle.scala

```diff
@@ -19,12 +21,15 @@ import com.digitalasset.daml.lf.value.Value.AbsoluteContractId
 import com.digitalasset.daml.lf.language.Ast.Package
 import com.digitalasset.daml.lf.data.Ref
 
-import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.{ExecutionContext, Future, Promise}
 
 /**
   * provides smart constructor for DAML engine environment
   */
 object SandboxDamle {
+  // Concurrent map of promises to request each package only once.
+  private[this] val packagePromises: ConcurrentHashMap[Ref.PackageId, Promise[Option[Package]]] =
```

Comment:
hey, the usage pattern seems to be an async read-through cache.  You might want to look at [scaffeine](https://github.com/blemale/scaffeine).

Reply:
hi @gabor-aranyossy! thanks for the suggestion. @gerolf-da would that library make more sense here?

Reply:
I wouldn't bring it in for this particular use case. But we have many others were using scaffeine could make sense. Thanks @gabor-aranyossy.